carma_cooperative_perception/src/msg_conversion.cpp

```diff
@@ -265,4 +382,150 @@ auto to_external_object_list_msg(
+  if (external_object.presence_vector & external_object.VELOCITY_INST_PRESENCE_VECTOR){
+    detected_object_common_data.speed.speed = std::hypot(external_object.velocity_inst.twist.linear.x, external_object.velocity_inst.twist.linear.y);
+
+    detected_object_common_data.presence_vector |= carma_v2x_msgs::msg::DetectedObjectCommonData::HAS_SPEED_Z;
+    detected_object_common_data.speed_z.speed = external_object.velocity_inst.twist.linear.z;
+
+    // TODO: heading - convert ang vel to scale heading
+    lanelet::BasicPoint3d external_object_position{external_object.pose.pose.position.x, external_object.pose.pose.position.y,external_object.pose.pose.position.z};
+    detected_object_common_data.heading.heading = remove_units(enu_orientation_to_true_heading(external_object.velocity_inst.twist.angular.z, external_object_position, map_projection));
+    
+  }
+
+  // optional data (determine based on object type)
+  // use object type struct for better control
+  carma_v2x_msgs::msg::DetectedObjectOptionalData detected_object_optional_data;
+
+  switch (external_object.object_type) {
+    case external_object.SMALL_VEHICLE:
+      detected_object_common_data.obj_type.object_type = j3224_v2x_msgs::msg::ObjectType::VEHICLE;
+
+      if(external_object.presence_vector & external_object.SIZE_PRESENCE_VECTOR){
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_SIZE;
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_HEIGHT;
+
+        detected_object_optional_data.det_veh.size.vehicle_width = external_object.size.y;
+        detected_object_optional_data.det_veh.size.vehicle_length = external_object.size.x;
+        detected_object_optional_data.det_veh.height.vehicle_height = external_object.size.z;
+      }
+      break;
+    case external_object.LARGE_VEHICLE:
+      detected_object_common_data.obj_type.object_type = j3224_v2x_msgs::msg::ObjectType::VEHICLE;
+
+      if(external_object.presence_vector & external_object.SIZE_PRESENCE_VECTOR){
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_SIZE;
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_HEIGHT;
+
+        detected_object_optional_data.det_veh.size.vehicle_width = external_object.size.y;
+        detected_object_optional_data.det_veh.size.vehicle_length = external_object.size.x;
+        detected_object_optional_data.det_veh.height.vehicle_height = external_object.size.z;
+      }
+      break;
+    case external_object.MOTORCYCLE:
+      detected_object_common_data.obj_type.object_type = j3224_v2x_msgs::msg::ObjectType::VEHICLE;
+
+      if(external_object.presence_vector & external_object.SIZE_PRESENCE_VECTOR){
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_SIZE;
```

Comment:
The `presence_vector` should be zeroed out initially to avoid undefined initial value issues.